controllers/mentor.go

```diff
@@ -0,0 +1,124 @@
+package controllers
+
+import (
+	"encoding/json"
+	"fmt"
+	"kwoc-backend/middleware"
+	"kwoc-backend/utils"
+	"net/http"
+
+	"github.com/rs/zerolog/log"
+	"gorm.io/gorm"
+	"kwoc-backend/models"
+)
+
+type RegisterMentorReqFields struct {
+	Username string `json:"username"`
+	Name     string `json:"name"`
+	Email    string `json:"email"`
+}
+
+func RegisterMentor(w http.ResponseWriter, r *http.Request) {
+	var reqFields = RegisterMentorReqFields{}
+
+	err := json.NewDecoder(r.Body).Decode(&reqFields)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprint(w, "Error decoding JSON body.")
+
+		log.Err(err).Msg("Error decoding JSON body.")
+		return
+	}
+
+	// Check if the JWT login username is the same as the mentor's given username
+	login_username := r.Context().Value(middleware.LoginCtxKey(middleware.LOGIN_CTX_USERNAME_KEY)).(string)
+
+	if reqFields.Username != login_username {
+		log.Warn().Msgf(
+			"%s %s %s\n%s %s",
+			r.Method,
+			r.RequestURI,
+			"POSSIBLE SESSION HIJACKING.",
+			fmt.Sprintf("JWT Username: %s", login_username),
+			fmt.Sprintf("Given Username: %s", reqFields.Username),
+		)
+
+		w.WriteHeader(http.StatusUnauthorized)
+		fmt.Fprint(w, "Login username and given username do not match.")
+		return
+	}
+
+	db, err := utils.GetDB()
+	if err != nil {
+		log.Err(err).Msgf(
+			"%s %s %s",
+			r.Method,
+			r.RequestURI,
+			"Error connecting to database.",
+		)
+
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Error connecting to database.")
+		return
+	}
```

Comment:
Bug here that needs refactoring asap. Two things to always keep in mind:
- Always add a `defer db.Close() ` after opening a database connection
- never create a db connection for each api hit. This is mainly due to two reasons:
 - adds unneeded latency to the controller.
 - can overload the database with high number of open connections. There should be a single global database connection persisted within the app ensuring the database never hits the max_connections irrespective of the qps on the server.
 
 
More on how to handle database connections and modelling can be read here: http://go-database-sql.org/

Reply:
> 

`db.Close()` doesn't exist in newer version of gorm (idk how it handles closing)

Reply:
Resolved in new commit (hopefully)